### ISSUE TITLE
how to make backwards compatible changes ?

### ISSUE THREAD

**Issue by ebuchman** [tendermint/go-amino#143]
The naiive thing doesn't seem to work:

```
package main

import (
	"fmt"
	"github.com/tendermint/go-amino"
)

type V1 struct {
	String string `json:"string"`
}

type V2 struct {
	String string `json:"string"`
	Int    int    `json:"int"`
}

func main() {
	cdc := amino.NewCodec()

	v2 := V2{String: "hi"}
	bz, _ := cdc.MarshalBinaryBare(v2)
	var v1 V1
	err := cdc.UnmarshalBinaryBare(bz, &v1)
	fmt.Println(err)
	fmt.Println(v1)

}
```

Returns:

```
Invalid typ3 byte
{hi}
```

So it does unmarshal whatever it can, but it also returns an error.

If the order of String and Int are reversed in V2, it just fails outright.

**Comment by liamsi** [comment]
Wouldn't this be simplest way to tackle this: 
when we read all fields but the buffer `bz` contains more then just the StructTerm, we continue reading / scanning (similar to what amino scan does). With that we just check that what is left in `bz` is validly decodable but we ignore everything we read until we read the last StructTerm and the buffer bz is fully consumed. Any reason this shouldn't work (@jaekwon)?

If it's not clear what this means, I can open a draftish PR. Should be quickly done.

> If the order of String and Int are reversed in V2, it just fails outright.

This wouldn't be solved by this approach though...

**Comment by liamsi** [comment]
I propose the following: 

First, we do as above to have backwards compatibility for when fields are added (without changes in the order).

Second, in a separate issue / PR we think about this changing the order thing. (Not sure this is possible, though.)

Thoughts?

**Comment by liamsi** [comment]
See: https://github.com/tendermint/go-amino/compare/develop...Liamsi:backwards_compat

**Comment by ebuchman** [comment]
I think that sounds right. Are amino fields numbered? I think @jaekwon was suggesting we add numbering.

The order switching isn't necessarily important to me, I was just pointing out the behaviour in comparison.

**Comment by mdyring** [comment]
By adding numbered fields, isn't this basically Protobuf again?

For the purpose of Tendermint/Cosmos, I first of all have to disagree with the use of Amino and storing type information for each and every object (details [here](https://github.com/tendermint/tendermint/issues/1524#issuecomment-389901372)). But I guess that ship has sailed.

Apart from adding numbered fields, it seems a natural way to achieve forward/backward compatibility would be to introduce a version number alongside the type information. This version number can then be considered at the (de)serialization points to determine how the object is populated based on the serialized data.

Some inspiration here: https://www.boost.org/doc/libs/1_60_0/libs/serialization/doc/tutorial.html#versioning (see "Class Versioning" and "Splitting serialize into save/load").

**Comment by liamsi** [comment]
Thanks for your feedback @ebuchman! :-)

> I think that sounds right. Are amino fields numbered? I think @jaekwon was suggesting we add numbering.

They are already numbered in the following sense:

> As in Protobuf, each struct field is keyed by a uvarint with the value (field_number << 3) | type, where type is 3 bits long. [¹⁾](https://github.com/tendermint/go-amino#structs-1)

In your example this numbering is deduced from the field order in the struct while encoding (using reflection). 

> The order switching isn't necessarily important to me, I was just pointing out the behaviour in comparison.

I'll clean up the code and add some tests for above approach and ask @jaekwon for a review. 

¹⁾ https://github.com/tendermint/go-amino#structs-1

**Comment by liamsi** [comment]
> By adding numbered fields, isn't this basically Protobuf again?

IMHO, It is conceptually already very similar to protobuf with some differences (as described in the readme).

> Some inspiration here: [...]

Thanks for the pointer @mdyring! I didn't know about this approach in boost. I like the idea of versioning but I would rather avoid adding this on a per struct / type basis. In the grpc / protobuf world versioning is sometimes used, too. But on the API, not on the types/(de)serilization function: https://cloud.google.com/apis/design/versioning (example: https://github.com/google/keytransparency/tree/master/core/api) With this approach the whole service is versioned and not each struct separately. 

For several reasons, I'm in favour of this kind of versioning, if any. If we have backwards compatibility as above, we can build versioned services without explicitly versioning each type.

**Comment by amrali** [comment]
From past experience, designing a protocol with versioned types/messages complicates handler code design by requiring an entire abstraction layer just to group different message versions per each high level API function/verb (i.e., the new API function A can only work with msg v2 because it added field X to the msg). Ultimately it will become a need to slap a version number on the protocol that represents a bunch of specific message versions in order to be able to document the protocol coherently. I believe versioned types are more suitable for more complex protocols that would justify the extra headache.

Also why not just use protobuff? :)

**Comment by liamsi** [comment]
Thanks for your feedback too @amrali :-)
 
> Also why not just use protobuff? :)

There are several reasons to use amino (at least internally). e.g., support for go interfaces (later java-interfaces and rust traits, too), no need to explicitly write in an intermediate message format (which currently speeds up development and doesn't create long verbose code), performance (ok, we need numbers here), and more (see also: https://github.com/tendermint/go-amino#amino-vs-protobuf3). I agree that we should support protobuf for non-internal modules though (due to protobufs wide adoption etc.).

**Comment by jaekwon** [comment]
I made review feedback for this PR: https://github.com/tendermint/go-amino/pull/155#pullrequestreview-122444439.

**Comment by liamsi** [comment]
basic backwards compatibility is resolved by #155
